Lib/test/test_traceback.py

```diff
@@ -2800,6 +3197,34 @@ def test_all(self):
                 expected.add(name)
         self.assertCountEqual(traceback.__all__, expected)
 
+    def test_levenshtein_distance(self):
+        # copied from _testinternalcapi.test_edit_cost
+        # to also exercise the Python implementation
+
+        def CHECK(a, b, expected):
+            actual = traceback._levenshtein_distance(a, b, 4044)
```

Comment:
This doesn't currently exercise the max_distance short-circuit paths very much, whereas _testinternalcapi.test_edit_cost does.

Could we add that and then make this same test code test both implementations?

Reply:
This is a rough draft of a possible randomized test case:


```python
    def test_levenshtein_distance_random(self):
        from functools import cache
        from traceback import _substitution_cost, _MOVE_COST

        @cache
        def levenshtein(a, b):
            if not a or not b:
                return (len(a) + len(b)) * _MOVE_COST
            option1 = levenshtein(a[:-1], b[:-1]) + _substitution_cost(a[-1], b[-1])
            option2 = levenshtein(a[:-1], b) + _MOVE_COST
            option3 = levenshtein(a, b[:-1]) + _MOVE_COST
            return min(option1, option2, option3)

        from random import choices, randrange

        for _ in range(1000):
            a = ''.join(choices("abcABC", k=randrange(10)))
            b = ''.join(choices("abcABC", k=randrange(10)))
            expected = levenshtein(a, b)
            res1 = traceback._levenshtein_distance(a, b, 1000)
            self.assertEqual(res1, expected, msg=(a, b))

            for threshold in [expected, expected + 1, expected + 2]:
                # big enough thresholds shouldn't change the result
                res2 = traceback._levenshtein_distance(a, b, threshold)
                self.assertEqual(res2, expected, msg=(a, b, threshold))

            for threshold in range(expected):
                # for small thresholds, the only piece of information
                # we receive is "strings not close enough".
                res3 = traceback._levenshtein_distance(a, b, threshold)
                self.assertGreater(res3, threshold, msg=(a, b, threshold))
```

Reply:
We talked about this at the sprint and a randomized test case feels somewhat icky as it might be unpredictable in CI. I'll put a predictable version of this instead.